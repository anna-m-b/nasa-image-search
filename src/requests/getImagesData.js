import axios from 'axios';

const getImagesData = async (searchInput, pageNumber = 1) => {
  const response = await axios.get(
    `https://images-api.nasa.gov/search?media_type=image&q=${searchInput}&page=${pageNumber}`
  );

  const collection = response.data.collection;

  const getAssets = async () => {
    return Promise.all(collection.items.map((item) => axios.get(item.href)));
  };

  const assetResponses = await getAssets();

  const imgLinks = assetResponses.map((asset) => asset.data);

  const results = collection.items.map((item) => {
    const id = new RegExp(item.data[0].nasa_id);
    const links = imgLinks.find((array) => id.test(array[0]));
    return {
      nasa_id: item.data[0].nasa_id,
      title: item.data[0].title,
      links,
    };
  });

  return results;
};

export default getImagesData;
