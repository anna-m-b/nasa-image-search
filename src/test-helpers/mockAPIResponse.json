{
  "data": {
    "collection": {
      "href": "https://images-api.nasa.gov/search?media_type=image&q=moon",
      "items": [
        {
            "href": "https://images-assets.nasa.gov/image/PIA12235/collection.json",
            "links": [
                {
                    "href": "https://images-assets.nasa.gov/image/PIA12235/PIA12235~thumb.jpg",
                    "render": "image",
                    "rel": "preview"
                }
            ],
            "data": [
                {
                    "center": "JPL",
                    "media_type": "image",
                    "description": "Nearside of the Moon",
                    "nasa_id": "PIA12235",
                    "secondary_creator": "ISRO/NASA/JPL-Caltech/Brown Univ.",
                    "keywords": [
                        "Moon",
                        "Chandrayaan-1"
                    ],
                    "description_508": "Nearside of the Moon",
                    "title": "Nearside of the Moon",
                    "date_created": "2009-09-24T18:00:22Z"
                }
            ]
        },
        {
            "href": "https://images-assets.nasa.gov/image/PIA13517/collection.json",
            "links": [
                {
                    "href": "https://images-assets.nasa.gov/image/PIA13517/PIA13517~thumb.jpg",
                    "render": "image",
                    "rel": "preview"
                }
            ],
            "data": [
                {
                    "center": "JPL",
                    "media_type": "image",
                    "description": "Color of the Moon",
                    "nasa_id": "PIA13517",
                    "secondary_creator": "NASA/GSFC/Arizona State University",
                    "keywords": [
                        "Moon",
                        "Lunar Reconnaissance Orbiter LRO"
                    ],
                    "description_508": "Color of the Moon",
                    "title": "Color of the Moon",
                    "date_created": "2010-09-10T22:24:40Z"
                }
            ]
        },
        {
            "href": "https://images-assets.nasa.gov/image/PIA12228/collection.json",
            "links": [
                {
                    "href": "https://images-assets.nasa.gov/image/PIA12228/PIA12228~thumb.jpg",
                    "render": "image",
                    "rel": "preview"
                }
            ],
            "data": [
                {
                    "center": "JPL",
                    "media_type": "image",
                    "description": "Cassini Look at Water on the Moon",
                    "nasa_id": "PIA12228",
                    "secondary_creator": "NASA/JPL-Caltech/USGS",
                    "keywords": [
                        "Moon",
                        "Cassini-Huygens"
                    ],
                    "description_508": "Cassini Look at Water on the Moon",
                    "title": "Cassini Look at Water on the Moon",
                    "date_created": "2009-09-24T18:00:15Z"
                }
            ]
        },
        {
            "href": "https://images-assets.nasa.gov/image/PIA12229/collection.json",
            "links": [
                {
                    "href": "https://images-assets.nasa.gov/image/PIA12229/PIA12229~thumb.jpg",
                    "render": "image",
                    "rel": "preview"
                }
            ],
            "data": [
                {
                    "center": "JPL",
                    "media_type": "image",
                    "description": "Mineral Mapping the Moon",
                    "nasa_id": "PIA12229",
                    "secondary_creator": "ISRO/NASA/JPL-Caltech/Brown Univ.",
                    "keywords": [
                        "Moon",
                        "Chandrayaan-1"
                    ],
                    "description_508": "Mineral Mapping the Moon",
                    "title": "Mineral Mapping the Moon",
                    "date_created": "2009-09-24T18:00:16Z"
                }
            ]
        },
        {
            "href": "https://images-assets.nasa.gov/image/PIA12233/collection.json",
            "links": [
                {
                    "href": "https://images-assets.nasa.gov/image/PIA12233/PIA12233~thumb.jpg",
                    "render": "image",
                    "rel": "preview"
                }
            ],
            "data": [
                {
                    "center": "JPL",
                    "media_type": "image",
                    "description": "Mapping the Moon, Point by Point",
                    "nasa_id": "PIA12233",
                    "secondary_creator": "NASA/JPL-Caltech",
                    "keywords": [
                        "Moon",
                        "Chandrayaan-1"
                    ],
                    "description_508": "Mapping the Moon, Point by Point",
                    "title": "Mapping the Moon, Point by Point",
                    "date_created": "2009-09-24T18:00:20Z"
                }
            ]
        }
      ],
      "version": "1.0",
      "metadata": {
        "total_hits": 9941
      },
      "links": [
        {
          "href": "https://images-api.nasa.gov/search?media_type=image&q=moon&page=2",
          "rel": "next",
          "prompt": "Next"
        }
      ]
    }
  }
}
