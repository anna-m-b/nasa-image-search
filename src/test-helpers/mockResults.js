export const mockResults = [
  {
    nasa_id: 'PIA18906',
    title: 'Sun Shines in High-Energy X-rays',
    links: [
      'http://images-assets.nasa.gov/image/PIA18906/PIA18906~orig.jpg',
      'http://images-assets.nasa.gov/image/PIA18906/PIA18906~medium.jpg',
      'http://images-assets.nasa.gov/image/PIA18906/PIA18906~small.jpg',
      'http://images-assets.nasa.gov/image/PIA18906/PIA18906~thumb.jpg',
      'http://images-assets.nasa.gov/image/PIA18906/metadata.json',
    ],
  },
  {
    nasa_id: 'PIA15179',
    title: 'The Sun Magnetic Field',
    links: [
      'http://images-assets.nasa.gov/image/PIA15179/PIA15179~orig.jpg',
      'http://images-assets.nasa.gov/image/PIA15179/PIA15179~small.jpg',
      'http://images-assets.nasa.gov/image/PIA15179/PIA15179~thumb.jpg',
      'http://images-assets.nasa.gov/image/PIA15179/metadata.json',
    ],
  },
  {
    nasa_id: 'PIA14729',
    title: 'Mercury Transit Across the Sun',
    links: [
      'http://images-assets.nasa.gov/image/PIA14729/PIA14729~orig.jpg',
      'http://images-assets.nasa.gov/image/PIA14729/PIA14729~thumb.jpg',
      'http://images-assets.nasa.gov/image/PIA14729/metadata.json',
    ],
  },
  {
    nasa_id: 'PIA17669',
    title: 'Pulses from the Sun',
    links: [
      'http://images-assets.nasa.gov/image/PIA17669/PIA17669~orig.jpg',
      'http://images-assets.nasa.gov/image/PIA17669/PIA17669~small.jpg',
      'http://images-assets.nasa.gov/image/PIA17669/PIA17669~thumb.jpg',
      'http://images-assets.nasa.gov/image/PIA17669/metadata.json',
    ],
  },
  {
    nasa_id: 'PIA09771',
    title: 'Filtering the Sun',
    links: [
      'http://images-assets.nasa.gov/image/PIA09771/PIA09771~orig.jpg',
      'http://images-assets.nasa.gov/image/PIA09771/PIA09771~thumb.jpg',
      'http://images-assets.nasa.gov/image/PIA09771/metadata.json',
    ],
  }
];
