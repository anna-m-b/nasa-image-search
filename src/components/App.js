import React, { useState } from 'react';
import '../styles/App.css';
import Header from './Header';
import SearchForm from './SearchForm';
import Results from './Results';
import getImagesData from '../requests/getImagesData';
import Navigation from './Navigation';
import Loading from './Loading';

function App() {
  const [searchInput, setSearchInput] = useState('');
  const [searchResults, setSearchResults] = useState();
  const [pageNumber, setPageNumber] = useState(0);
  const [isLoading, setIsLoading] = useState(false);
  const [isError, setIsError] = useState(false);

  const onChange = (e) => {
    setSearchInput(e.target.value);
  };

  const onSearchSubmit = async (e) => {
    e.preventDefault();
    setIsLoading(true);
    setSearchResults();
    try {
      const results = await getImagesData(searchInput);
      setSearchResults(results);
      setPageNumber(1);
    } catch (error) {
      setIsError(true);
    }
    setIsLoading(false);
  };

  const loadNextPage = async () => {
    const results = await getImagesData(searchInput, pageNumber + 1);
    setSearchResults(results);
    setPageNumber((prevPage) => prevPage + 1);
  };

  const loadPrevPage = async () => {
    const results = await getImagesData(searchInput, pageNumber - 1);
    setSearchResults(results);
    setPageNumber((prevPage) => prevPage - 1);
  };

  return (
    <div className="App">
      <Header />
      <SearchForm
        onChange={onChange}
        searchInput={searchInput}
        onSearchSubmit={onSearchSubmit}
      />
      {isLoading && <Loading />}
      {isError && (
        <p className="error-message">
          Oops, something went wrong. Please try again!
        </p>
      )}
      {searchResults && !searchResults.length && (
        <p className="no-results-message">
          Oh no! Nothing was found in the database.
        </p>
      )}
      {searchResults && searchResults.length > 0 && (
        <Results results={searchResults} />
      )}
      <Navigation
        loadNextPage={loadNextPage}
        loadPrevPage={loadPrevPage}
        pageNumber={pageNumber}
      />
    </div>
  );
}

export default App;
