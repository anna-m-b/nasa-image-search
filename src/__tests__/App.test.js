import React from 'react';
import { render, waitFor, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from '../components/App';
import getImagesData from '../requests/getImagesData';
import { mockResults } from '../test-helpers/mockResults';

jest.mock('../requests/getImagesData');
jest.mock('../components/Results.js', () => {
  return {
    __esModule: true,
    default: () => {
      return <div className="results"></div>;
    },
  };
});

describe('App', () => {
  beforeEach(() => {
    render(<App />);
  });

  describe('on loading, before searching', () => {
    test('renders correctly', () => {
      const { asFragment } = render(<App />);
      expect(asFragment()).toMatchSnapshot();
    });

    test('captures user input', async () => {
      const input = screen.getByRole('textbox');
      userEvent.type(input, 'sun');
      await waitFor(() => {
        expect(input).toHaveValue('sun');
      });
    });

    test('calls the getImagesData function with user input', async () => {
      const input = screen.getByRole('textbox');
      const button = screen.getByRole('button', { name: /search/i });
      userEvent.type(input, 'sun');
      userEvent.click(button);
      await waitFor(() => {
        expect(getImagesData).toHaveBeenCalled();
        expect(getImagesData).toHaveBeenCalledWith('sun');
      });
    });
  });

  describe('with search result', () => {
    test('loading text displays while images load', async () => {
      userEvent.click(screen.getByRole('button', { name: /search/i }));
      await waitFor(() => {
        expect(screen.getByText(/loading/i)).toBeInTheDocument();
      });
    });

    test('loading text is hidden after app receives response from api', async () => {
      userEvent.click(screen.getByRole('button', { name: /search/i }));
      getImagesData.mockResolvedValue(mockResults);
      await waitFor(() => {
        expect(screen.queryByText(/loading/i)).not.toBeInTheDocument();
      });
    });

    test('an error message is displayed if the api responds with an error', async () => {
      getImagesData.mockResolvedValue(Promise.reject('failed'));
      userEvent.click(screen.getByRole('button', { name: /search/i }));
      await waitFor(() => {
        expect(
          screen.getByText(/Oops, something went wrong/i)
        ).toBeInTheDocument();
        expect(screen.getByText(/Oops, something went wrong/i)).toHaveClass(
          'error-message'
        );
      });
    });

    test('a message is displayed if there are no results', async () => {
      getImagesData.mockResolvedValue([]);
      userEvent.click(screen.getByRole('button', { name: /search/i }));
      await waitFor(() => {
        expect(
          screen.getByText(/Oh no! Nothing was found in the database/i)
        ).toBeInTheDocument();
        expect(
          screen.getByText(/Oh no! Nothing was found in the database/i)
        ).toHaveClass('no-results-message');
        screen.debug()
      });
    });
  });
});

// To Do

// check error handling behaviour
