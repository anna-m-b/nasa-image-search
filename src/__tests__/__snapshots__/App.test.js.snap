// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App on loading, before searching renders correctly 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="header"
    >
      <img
        alt="Nasa Logo"
        class="header_nasa-logo"
        src="nasa.jpg"
        width="500px"
      />
    </div>
    <form
      class="search-form"
    >
      <label
        class="search-form_label"
        for="search-input"
      >
        Search the NASA images database
      </label>
      <span
        class="search-form_input-and-button"
      >
        <input
          class="search-form_input"
          id="search-input"
          placeholder="Enter a search term"
          type="text"
          value=""
        />
        <button
          class="search-form_button"
          type="submit"
        >
          Search
        </button>
      </span>
    </form>
    <div
      class="navigation"
    >
      <button
        class="prev-page_button"
        disabled=""
      >
        Previous Page
      </button>
      <button
        class="next-page_button"
        disabled=""
      >
        Next Page
      </button>
    </div>
  </div>
</DocumentFragment>
`;
